src/System.Runtime.Extensions/tests/System/Runtime/CompilerServices/SwitchExpressionExceptionTests.cs

```diff
@@ -0,0 +1,47 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.CompilerServices;
+using Xunit;
+
+namespace System.Runtime.CompilerServices.Tests
+{
+    public class SwitchExpressionExceptionTests
+    {
+        [Fact]
+        public static void DefaultConstructor()
+        {
+            var ex = new SwitchExpressionException();
+
+            Assert.NotNull(ex.Message);
+        }
+
+        [Fact]
+        public static void Constructor_StringAsObjectArg()
+        {
+            string message = "stringInput";
+            var ex = new SwitchExpressionException(message);
```

Comment:
Usually string input on Exception would be the exception message. But note, as is, we now have just a default constructor and another one taking object. The later will be set as UnmatchedValue property and not as exception message.

Reply:
I agree that's strange.  No one should need to construct one of these manually, but still.  @terrajobst?

Reply:
I think what should have been approved is 
```c#
        public SwitchExpressionException() { }
        public SwitchExpressionException(string message) { }
        public SwitchExpressionException(object unmatchedValue) { }
        public SwitchExpressionException(string message, object unmatchedValue) { }
```

Precedent is `public ArgumentException(string message, string paramName)` and perversely `public ArgumentOutOfRangeException(string paramName, string message)`. Pick 😢

Reply:
I disagree.  If we want there to be a non-null `Exception.Message`, then the constructor `SwitchExpressionException( object unmatchedValue)` should pass a constant (or localized) string, like "The switch expression's input was not matched by any branch.", to its base constructor.  The compiler should not be forced to supply a message when it wants to provide an unmatched value.

Reply:
I don't feel strongly.

Reply:
The way I understand [the guideline](https://docs.microsoft.com/en-us/visualstudio/code-quality/ca1032-implement-standard-exception-constructors?view=vs-2017) is that there should (generally) always be four constructors:

* `public SwitchExpressionException()`
* `public SwitchExpressionException(string)`
* `public SwitchExpressionException(string, Exception)`
* `private SwitchExpressionException(SerializationInfo, StreamingContext)`

It doesn't preclude the exception from having additional constructors. The problem with the argument exception is that there are cases where the first string parameter isn't the message, which can be highly confusing.

In this case, @gafter is asking for this:

* `public SwitchExpressionException(object value)`

Which is fine, IMHO. It seems reasonable that we don't want the compiler to code spit a specific message or resource lookup. If we feel strongly that this shouldn't be a constructor I'm sure the Roslyn folks would be OK with a factory method too, such as:

* `static SwitchExpressionException ForValue(object value)`

I'd just add the additional constructor.

Reply:
Sounds good to me

Reply:
So just to confirm this is the final approved APIs?

```
    public sealed class SwitchExpressionException : System.InvalidOperationException
    {
        public SwitchExpressionException() { }
        public SwitchExpressionException(Exception innerException) { }
        public SwitchExpressionException(object unmatchedValue) { }
        public SwitchExpressionException(string message) { }
        public SwitchExpressionException(string message, Exception innerException) { }
        public object UnmatchedValue { get { throw null; } }
        public override void GetObjectData(System.Runtime.Serialization.SerializationInfo info, System.Runtime.Serialization.StreamingContext context) { }
    }

```

Reply:
Looks reasonable to me.